frame/vault/src/lib.rs

```diff
@@ -98,6 +103,97 @@ pub mod pallet {
 	pub type VaultInfo<T> =
 		crate::models::VaultInfo<AccountIdOf<T>, BalanceOf<T>, AssetIdOf<T>, BlockNumberOf<T>>;
 
+	pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"vaul");
+	pub const CRYPTO_KEY_TYPE: CryptoKeyTypeId = CryptoKeyTypeId(*b"vaul");
+
+	pub mod crypto {
+		use super::KEY_TYPE;
+		use sp_core::sr25519::Signature as Sr25519Signature;
+		use sp_runtime::{
+			app_crypto::{app_crypto, sr25519},
+			traits::Verify,
+			MultiSignature, MultiSigner,
+		};
+		app_crypto!(sr25519, KEY_TYPE);
+
+		pub struct TestAuthId;
+
+		// implementation for runtime
+		impl frame_system::offchain::AppCrypto<MultiSigner, MultiSignature> for TestAuthId {
+			type RuntimeAppPublic = Public;
+			type GenericSignature = sp_core::sr25519::Signature;
+			type GenericPublic = sp_core::sr25519::Public;
+		}
+
+		// implementation for mock runtime in test
+		impl
+			frame_system::offchain::AppCrypto<
+				<Sr25519Signature as Verify>::Signer,
+				Sr25519Signature,
+			> for TestAuthId
+		{
+			type RuntimeAppPublic = Public;
+			type GenericSignature = sp_core::sr25519::Signature;
+			type GenericPublic = sp_core::sr25519::Public;
+		}
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	pub trait Config: CreateSignedTransaction<Call<Self>> + frame_system::Config {
+		#[allow(missing_docs)]
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ SaturatingSub
+			+ AtLeast32BitUnsigned
+			+ From<u128>
+			+ Zero;
+		type CurrencyFactory: CurrencyFactory<Self::AssetId>;
+		type AssetId: FullCodec
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ From<u128>
+			+ TypeInfo;
+		type NativeCurrency: TransferNative<Self::AccountId, Balance = Self::Balance>
+			+ MutateNative<Self::AccountId, Balance = Self::Balance>
+			+ MutateHoldNative<Self::AccountId, Balance = Self::Balance>;
+		type Currency: Transfer<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>
+			+ Mutate<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>
+			+ MutateHold<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>;
+		type VaultId: AddAssign
+			+ FullCodec
+			+ One
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ Into<u128>
+			+ From<u64>
+			+ TypeInfo;
+		type Convert: Convert<Self::Balance, u128> + Convert<u128, Self::Balance>;
+		type MaxStrategies: Get<usize>;
+		type MinimumDeposit: Get<Self::Balance>;
+		type MinimumWithdrawal: Get<Self::Balance>;
+		type CreationDeposit: Get<Self::Balance>;
+		type ExistentialDeposit: Get<Self::Balance>;
+		type TombstoneDuration: Get<Self::BlockNumber>;
+		type RentPerBlock: Get<Self::Balance>;
+		type PalletId: Get<PalletId>;
+	}
+
+	#[cfg(not(feature = "runtime-benchmarks"))]
```

Comment:
It needs for an off-chain worker only.

Reply:
I think i still need it as I need more constraint such as Into/From for some types...

Reply:
Ok, but `crypto` module (109-139 lines) looks redundant.
`CreateSignedTransaction<Call<Self>>` needs for an off-chain worker only.

Reply:
I will remove all the tweaking done for the accountid of course.